zerver/migrations/0147_realm_emoji_drop_unique_constraint.py

```diff
@@ -0,0 +1,105 @@
+
+from boto.s3.bucket import Bucket
+from boto.s3.connection import S3Connection
+from boto.s3.key import Key
+from django.conf import settings
+from django.db import migrations, models
+from django.db.backends.postgresql_psycopg2.schema import DatabaseSchemaEditor
+from django.db.migrations.state import StateApps
+
+class Uploader:
+    def __init__(self) -> None:
+        self.old_orig_image_path_template = "{realm_id}/emoji/{emoji_file_name}.original"
+        self.old_path_template = "{realm_id}/emoji/{emoji_file_name}"
+        self.new_orig_image_path_template = "{realm_id}/emoji/images/{emoji_file_name}.original"
+        self.new_path_template = "{realm_id}/emoji/images/{emoji_file_name}"
+
+    def copy_files(self, src_path: str, dst_path: str) -> None:
+        raise NotImplementedError()
+
+    def ensure_emoji_images(self, realm_id: int, old_filename: str, new_filename: str) -> None:
+        # Copy original image file.
+        old_file_path = self.old_orig_image_path_template.format(realm_id=realm_id,
+                                                                 emoji_file_name=old_filename)
+        new_file_path = self.new_orig_image_path_template.format(realm_id=realm_id,
+                                                                 emoji_file_name=new_filename)
+        self.copy_files(old_file_path, new_file_path)
+
+        # Copy resized image file.
+        old_file_path = self.old_path_template.format(realm_id=realm_id,
+                                                      emoji_file_name=old_filename)
+        new_file_path = self.new_path_template.format(realm_id=realm_id,
+                                                      emoji_file_name=new_filename)
+        self.copy_files(old_file_path, new_file_path)
+
+class LocalUploader(Uploader):
+    def __init__(self) -> None:
+        super().__init__()
+
+    @staticmethod
+    def mkdirs(path: str) -> None:
+        dirname = os.path.dirname(path)
+        if not os.path.isdir(dirname):
+            os.makedirs(dirname)
+
+    def copy_files(self, src_path: str, dst_path: str) -> None:
+        src_path = os.path.join(settings.LOCAL_UPLOADS_DIR, 'avatars', src_path)
+        self.mkdirs(src_path)
+        dst_path = os.path.join(settings.LOCAL_UPLOADS_DIR, 'avatars', dst_path)
+        self.mkdirs(dst_path)
+        shutil.copyfile(src_path, dst_path)
+
+class S3Uploader(Uploader):
+    def __init__(self) -> None:
+        super().__init__()
+        conn = S3Connection(settings.S3_KEY, settings.S3_SECRET_KEY)
+        bucket_name = settings.S3_AVATAR_BUCKET
+        self.bucket = conn.get_bucket(bucket_name, validate=False)
+
+    def copy_files(self, src_key: str, dst_key: str) -> None:
+        self.bucket.copy_key(src_key, self.bucket, dst_key)
+
+def get_uploader() -> Uploader:
+    if settings.LOCAL_UPLOADS_DIR is None:
+        return S3Uploader()
+    return LocalUploader()
+
+def get_emoji_file_name(emoji_file_name: str, emoji_id: int) -> str:
```

Comment:
Shouldn't we be using Text as type annotations instead of str here? This goes for all the occurrences in the file.

Reply:
`Text` and `str` are the same thing in Python 3.  We're eventually going to migrate them all to `str`, but it's a pain.  So this is preferred.